server/hotrod/src/main/java/org/infinispan/server/hotrod/Decoder2x.java

```diff
@@ -332,32 +345,120 @@ public void customReadHeader(HotRodHeader header, ByteBuf buffer, CacheDecodeCon
                }
             }
             break;
+         case PREPARE_TX:
+            PrepareTransactionContext ctx = (PrepareTransactionContext) hrCtx.operationDecodeContext;
+            if (ctx == null) {
+               Optional<PrepareTransactionContext> optCtx = readPrepareTxContext(buffer);
+               if (optCtx.isPresent()) {
+                  ctx = optCtx.get();
+                  hrCtx.operationDecodeContext = ctx;
+                  buffer.markReaderIndex();
+                  if (trace) {
+                     log.tracef("Decoded Prepare Transaction Context: %s", ctx);
+                  }
+               } else {
+                  return;
+               }
+            }
+            while (!ctx.isFinished()) {
+               Optional<TransactionWrite> txWrite = readTransactionWrite(header, buffer);
+               if (txWrite.isPresent()) {
+                  TransactionWrite write = txWrite.get();
+                  ctx.addWrite(write);
+                  buffer.markReaderIndex();
+                  if (trace) {
+                     log.tracef("Decoded Transaction Write: %s", write);
+                  }
+               } else {
+                  return;
+               }
+            }
+            out.add(hrCtx);
+            break;
+         case COMMIT_TX:
+         case ROLLBACK_TX:
+            readXid(buffer).ifPresent(xid -> {
+               hrCtx.operationDecodeContext = xid;
+               out.add(hrCtx);
+               buffer.markReaderIndex();
+               if (trace) {
+                  log.tracef("Decoded commit/rollback XID: %s", xid);
+               }
+            });
+            break;
          default:
             // This operation doesn't need additional reads - has everything to process
             out.add(hrCtx);
       }
    }
 
+   private Optional<PrepareTransactionContext> readPrepareTxContext(ByteBuf buffer) {
+      return readXid(buffer)
+            .flatMap(xid -> readMaybeByte(buffer)
+                  .flatMap(onePhaseCommit -> readMaybeVInt(buffer)
+                        .map(numberOfWrites ->
+                              new PrepareTransactionContext(xid, onePhaseCommit == 1, numberOfWrites))));
+   }
+
+   private Optional<TransactionWrite> readTransactionWrite(HotRodHeader header, ByteBuf byteBuf) {
+      return readMaybeRangedBytes(byteBuf).flatMap(key ->
```

Comment:
I don't like it either but it seems the way to avoid blocking the reader thread for data.